compiler/rustc_ast_lowering/src/block.rs

```diff
@@ -159,6 +159,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             span,
             kind: hir::ExprKind::If(let_expr, then_expr, Some(else_expr)),
         });
+        let drop_temps = self.expr_drop_temps(span, if_expr, AttrVec::new());
```

Comment:
to be consistent with `let` we would have to actually drop temps for the `let_expr` or `init_expr`, wouldn't we?

Reply:
Yeah, gut feeling this would fix the new problem. Good idea.

Reply:
I imagine the most straightforward solution will drop temps before the `else` block instead of after it. Maybe that's okay, but just want to make sure that detail is considered.

Reply:
My proposed fix does not fix the general discrepancy between let and let-else, e.g https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=57e814fa57019bc62930bc9d61c894da

after the last lang meeting either @nikomatsakis or @pnkfelix should take over the review here. i am not familar enough with rusts temporary rules

Reply:
Should a new issue be filed? Or work it within this issue